### PR TITLE
DPC++/OpenCL interoperability sample

### PR DESCRIPTION
**Pull request by karlqi2000**
# Adding New samples
## Description
The samples here show how to incrementally migrate OpenCL to DPC++.
1. Shows a DPC++ program that compiles and runs an OpenCL kernel. (For this, OpenCL must be set as the backend and not Level 0)
    - Use the environment variable SYCL_DEVICE_FILTER=OPENCL
3. Show how to convert OpenCL objects (Memory Objects, Platform, Context, Program, Kernel) to DPC++ and execute the program. 

### Administrative
- [X] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [ ]  Sample has dependencies and/or Third Party binaries. If yes inform the oneAPI Code Samples Project Manager: @JoeOster

### Code Development
- [X] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [X] Adhere to readme template 
- [X] Enforce format via clang-format config file
- [X] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [X] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [X] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

### Security and Legal
- [X] **_N/A -_** OSPDT Approval
- [X] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [X] Protex (see @JoeOster for assistance)
- [X] IPLDT Scan (see @JoeOster for assistance) Approved 05/18/2021
- [X] Virus scan

### Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Joe Oster(@JoeOster)
- [X] Tested using Dev Cloud when applicable

### CODE REVIEW

**Review by JoeOster — CHANGES_REQUESTED** [review]
The readme  needs to be expanded with the following subjects, please see the mandelbrot sample for an example
Introduction
Optimize for table
OS/Hardware & software needed
What you will learn
Use cases covered
Time to complete
Detailed explanation or link to a detailed explanation with infographics

**Comment by karlqi2000** [comment]
I believe the requested change is implemented. Let me know if anything else is needed.

**Comment by tomlenth** [comment]
The readme is still missing quite a bit.  Please see https://github.com/oneapi-src/oneAPI-samples/blob/master/DirectProgramming/DPC%2B%2B/CombinationalLogic/mandelbrot/README.md for an example of all the needed sections as well as example of the level of detail required.   you may want to move some of the intro material into the Purpose, then expand the concepts quite a bit more.  you really want to ask why would someone want to use this and why they'd want to explore interoperability with dpc++. for example, at first glance, I didn't get that it's really for how to migrate incrementally.  It would be good to show that benefit and why.  

Also, see the additional sections in the readme sample using mandelbrot above.  Your readme should have all of those sections.  Feel free to reach out if you want to walk through these or talk about the purpose.

**Review by JoeOster — CHANGES_REQUESTED** [review]

**Comment by karlqi2000** [comment]
I've change the sample to use cmake and also beefed up the README

**Review by JoeOster — DISMISSED** [review]
approving for ci

**Comment by tomlenth** [comment]
The sample readme is still missing quite a bit from the template.  Take a look at the mandelbrot readme.  your readme should have all of the sections, including devcloud, sample output, etc.  pleaes try to make it look and feel as consistent with that readme as possible.

**Comment by karlqi2000** [comment]
I've added more details to the README

File: DirectProgramming/DPC++/OpenCLInterop/README.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+> Note: if you have not already done so, set up your CLI 
+> environment by sourcing  the setvars script located in 
+> the root of your oneAPI installation. 
+>
+> Linux Sudo: . /opt/intel/oneapi/setvars.sh  
+> Linux User: . ~/intel/oneapi/setvars.sh  
+> Windows: C:\Program Files(x86)\Intel\oneAPI\setvars.bat
+
+### Running Samples In DevCloud
+If running a sample in the Intel DevCloud, remember that you must specify the compute node (CPU, GPU, FPGA) and whether to run in batch or interactive mode. For more information, see the Intel® oneAPI Base Toolkit Get Started Guide (https://devcloud.intel.com/oneapi/get-started/base-toolkit/)
```
Comment (JoeOster) [inline DirectProgramming/DPC++/OpenCLInterop/README.md:41]: I believe the correct link would be https://devcloud.intel.com/oneapi/get_started/baseToolkitSamples/

**Review by JoeOster — CHANGES_REQUESTED** [review]

File: DirectProgramming/DPC++/OpenCLInterop/README.md (hunk 1 above)
Comment (karlqi2000) [reply DirectProgramming/DPC++/OpenCLInterop/README.md]: Yes, this would be the better URL. I've made the change.

**Review by tomlenth — CHANGES_REQUESTED** [review]
Please add the section about the include file. should follow immediately after the env variables. 

Also add a top level #Haeading "Running the Sample" before the example output.

**Comment by karlqi2000** [comment]
Tom, there is no special include files needed for this sample, just the standard sycl and OpenCL ones. There are also no arguments for the executable, so the instructions with the make run* already covers all the execution scenarios. That's why I didn't think it was necessary to add those sections.

File: DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp [hunk 2]
```diff
@@ -0,0 +1,127 @@
+//==============================================================
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+// =============================================================
+
+#include <stdio.h>
+#include <CL/sycl.hpp>
+#include "CL/opencl.h"
+using namespace sycl;
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp:10]: Please verify this was run through clang-format.

File: DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+#define MAX_SOURCE_SIZE (0x100000)
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp:12]: This should be constexpr not #define

File: DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+static const int N = 1024;
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp:13]: This should be constexpr not static const

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

File: DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp (hunk 3 above)
Comment (karlqi2000) [reply DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp]: Done

File: DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp (hunk 2 above)
Comment (karlqi2000) [reply DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp:10]: Yes this was run through clang-format

**Review by JoeOster — DISMISSED** [review]
approving for ci only

File: DirectProgramming/DPC++/OpenCLInterop/README.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
+# DPC++ OpenCL&trade; Interoperability Example
+
+The examples here demonstrate how DPC++ can interact with OpenCL&trade;. This enables programmers to incrementally migrate from
+OpenCL to DPC++. Two usage scenarios are shown. First is a DPC++ program that compiles and runs an OpenCL kernel. The second program converts OpenCL objects to DPC++.
+
+For more information on migrating from OpenCL to DPC++, see [Migrating OpenCL Designs to DPC++](https://software.intel.com/content/www/us/en/develop/articles/migrating-opencl-designs-to-dpcpp.html).
+
+| Optimized for                       | Description
+|:---                               |:---
+| OS                                | Linux* Ubuntu* 18.04, 20; Windows 10
```
Comment (JoeOster) [inline DirectProgramming/DPC++/OpenCLInterop/README.md:10]: Why does this sample show as working on windows, as I am not finding instructions in readme and json is linux only?

Comment (karlqi2000) [reply DirectProgramming/DPC++/OpenCLInterop/README.md]: That's just an artifact of copying. Although the code itself should run on windows as well. I've removed windows 10 from README

**Review by JoeOster — DISMISSED** [review]
approved for ci only

File: DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_kernel.dp.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                                int i=get_global_id(0);
+                                c[i] = a[i] + b[i]; 
+                            } )");
+
+  buffer<int, 1> buf_a(a, range<1>(N));
+  buffer<int, 1> buf_b(b, range<1>(N));
+  buffer<int, 1> buf_c(c, range<1>(N));
+
+  q.submit([&](handler &h) {
+    auto A = buf_a.get_access<access::mode::read>(h);
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_kernel.dp.cpp:35]: Should use SYCL 2020 accessors.

File: DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    std::cout << "Device: "
+              << sycl_queue.get_device().get_info<info::device::name>()
+              << std::endl;
+    kernel sycl_kernel(ocl_kernel, sycl_context);
+    buffer<int, 1> sycl_buf_a(ocl_buf_a, sycl_context);
+    buffer<int, 1> sycl_buf_b(ocl_buf_b, sycl_context);
+    buffer<int, 1> sycl_buf_c(ocl_buf_c, sycl_context);
+    sycl_queue.submit([&](handler &h) {
+      // Create accessors for each of the buffers
+      auto a_accessor = sycl_buf_a.get_access<access::mode::read>(h);
```
Comment (pmpeter1) [inline DirectProgramming/DPC++/OpenCLInterop/src/dpcpp_with_opencl_objects.dp.cpp:95]: Should use SYCL 2020 accessors.

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

**Review by JoeOster — DISMISSED** [review]

**Review by pmpeter1 — APPROVED** [review]

File: DirectProgramming/DPC++/OpenCLInterop/README.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
+# DPC++ OpenCL&trade; Interoperability Example
+
+The examples here demonstrate how DPC++ can interact with OpenCL&trade;. This enables programmers to incrementally migrate from
```
Comment (JoeOster) [inline DirectProgramming/DPC++/OpenCLInterop/README.md]: This intro paragraph needs to flow better.

**Review by JoeOster — CHANGES_REQUESTED** [review]

**Review by JoeOster — APPROVED** [review]

**Pull request merged**